discovery-provider/src/tasks/index_aggregate_plays.py

```diff
@@ -1,17 +1,103 @@
 import logging
 import time
+import sqlalchemy as sa
 from src.tasks.celery_app import celery
+from src.models import LatestSlots, Play
 
 logger = logging.getLogger(__name__)
 
+AGGREGATE_PLAYS_TABLE_NAME = "aggregate_plays"
 
-def update(self, db):
-    with db.scoped_session() as session:
-        start_time = time.time()
-        session.execute("REFRESH MATERIALIZED VIEW CONCURRENTLY aggregate_plays")
+### UPDATE_AGGREGATE_PLAYS_QUERY ###
+# Get new plays that came after the previous most recent indexed slot for aggregate_play
+# Group those new plays by play item id to get the aggregate play counts
+# For new play item ids, insert those aggregate counts
+# For existing play item ids, add the new aggregate count to the existing aggregate count
+UPDATE_AGGREGATE_PLAYS_QUERY = """
+    WITH aggregate_plays_latest_slot AS (
+        SELECT
+            :prev_latest_slot AS prev_latest_slot,
+            :new_latest_slot AS new_latest_slot
+    ),
+    new_plays AS (
+        SELECT
+            play_item_id,
+            count(play_item_id) AS count
+        FROM
+            plays p
+        WHERE
+            p.slot > (
+                SELECT
+                    prev_latest_slot
+                FROM
+                    aggregate_plays_latest_slot
+            )
+            AND p.slot <= (
+                SELECT
+                    new_latest_slot
+                FROM
+                    aggregate_plays_latest_slot
+            )
+        GROUP BY
+            play_item_id
+    )
+    INSERT INTO
+        aggregate_plays (play_item_id, count)
+    SELECT
+        new_plays.play_item_id,
+        new_plays.count
+    FROM
+        new_plays ON CONFLICT (play_item_id) DO
+    UPDATE
+    SET
+        count = aggregate_plays.count + EXCLUDED.count
+    """
+
+UPSERT_LATEST_SLOTS = """
+    INSERT INTO latest_slots (tablename, slot)
+    VALUES(:tablename, :slot)
+    ON CONFLICT (tablename)
+    DO UPDATE SET slot = EXCLUDED.slot;
+    """
+
+def _update_aggregate_plays(session):
+    # get the last updated slot that counted towards the current aggregate plays
+    prev_latest_slot = (session.query(LatestSlots.slot)
+        .filter(LatestSlots.tablename == AGGREGATE_PLAYS_TABLE_NAME)
+    ).scalar()
+
+    if not prev_latest_slot:
+        prev_latest_slot = 0
+
+    # get the new latest
+    new_latest_slot = (
+        session.query(Play.slot)
+        .filter(Play.slot != None)
+        .order_by(Play.slot.desc()).limit(1)
+    ).scalar()
+
+    if not new_latest_slot:
+        new_latest_slot = 0
 
-    logger.info(
-        f"index_aggregate_plays.py | Finished updating in: {time.time() - start_time} sec."
+    # update aggregate plays with new plays that came after the prev_latest_slot
+    logger.info(f"index_aggregate_plays.py | Updating {AGGREGATE_PLAYS_TABLE_NAME}")
+
+    session.execute(
```

Comment:
just to confirm, since session is passed into this function, the two queries below happen in the same tx?

Reply:
Yup, everything in the context manager's scope is the same transaction. I also tested that by setting breakpoints and throw an exception in the transaction. If the tests had a mock db, we could test that behavior.